### PR TITLE
Guard L2C regions against corrupted data

### PR DESCRIPTION
**Pull request by umpirsky**
Fixes https://github.com/doctrine/orm/issues/7266.

### CODE REVIEW

**Comment by greg0ire** [comment]
Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

**Comment by umpirsky** [comment]
@greg0ire Improved, thanks.

**Comment by greg0ire** [comment]
Your now exemplary commit message says:

> On cache miss the parameter `$entityEntry` sometimes will be false. 

But apparently, `false` is not supposed to happen at all:

https://github.com/doctrine/orm/blob/a896421be2eb9f9b6afd2c358f725462c234cefb/lib/Doctrine/ORM/Cache/MultiGetRegion.php#L20

IMO the right fix would be to make sure you cannot get `false` instead of handling it afterwards. Make the implementation conform to what the contract specified in the phpdoc says.

**Comment by umpirsky** [comment]
I agree, this was a quick way for me to turn error into cache miss. But the root of the problem is in `getMultiple()` in deed, and I don't see the way to fix that.

**Comment by greg0ire** [comment]
> `getMultiple()`

Which implementation are we talking about here? Whatever it is, shouldn't we ensure it never returns `false` as an element of the returned array?

**Comment by umpirsky** [comment]
> Which implementation are we talking about here? Whatever it is, shouldn't we ensure it never returns `false` as an element of the returned array?

It's [RedisCache](https://github.com/doctrine/cache/blob/master/lib/Doctrine/Common/Cache/RedisCache.php). It looks like [this condition](https://github.com/doctrine/cache/blob/master/lib/Doctrine/Common/Cache/RedisCache.php#L60) can allow `false` values if `$this->redis->exists($key)` is `true`.

**Comment by greg0ire** [comment]
`RedisCache` does not implement `MultiGetRegion`, does it? We're looking for an implementation of `MultiGetRegion` that returns a collection of `CacheEntry` objects, and lets a `false` slip in.

**Comment by greg0ire** [comment]
I think the issue might be here: https://github.com/doctrine/orm/blob/2726636e55fdbdf6fb1ed466f22d34de116fb8f9/lib/Doctrine/ORM/Cache/Region/DefaultMultiGetRegion.php#L53-L55

See the difference with the other implementation: https://github.com/doctrine/orm/blob/2726636e55fdbdf6fb1ed466f22d34de116fb8f9/lib/Doctrine/ORM/Cache/Region/DefaultRegion.php#L85-L90

**Comment by umpirsky** [comment]
Yes, that looks wrong, it's worth trying.

**Comment by umpirsky** [comment]
I can't simulate it in the test case:

```diff
--- a/tests/Doctrine/Tests/ORM/Cache/MultiGetRegionTest.php
+++ b/tests/Doctrine/Tests/ORM/Cache/MultiGetRegionTest.php
@@ -38,4 +38,26 @@ class MultiGetRegionTest extends AbstractRegionTest
         self::assertEquals($value1, $actual[0]);
         self::assertEquals($value2, $actual[1]);
     }
+
+    public function testGetMultiCacheMiss() : void
+    {
+        $key1   = new CacheKeyMock('key.1');
+        $value1 = new CacheEntryMock(['id' => 1, 'name' => 'bar']);
+
+        $key2   = new CacheKeyMock('key.2');
+        $value2 = new CacheEntryMock(['id' => 2, 'name' => 'bar']);
+
+        self::assertFalse($this->region->contains($key1));
+        self::assertFalse($this->region->contains($key2));
+
+        $this->region->put($key1, $value1);
+        $this->region->put($key2, $value2);
+
+        self::assertTrue($this->region->contains($key1));
+        self::assertTrue($this->region->contains($key2));
+
+        $actual = $this->region->getMultiple(new CollectionCacheEntry([$key1, $key2]));
+
+        self::assertNull($actual);
+    }
```

I don't know how to create `CacheEntryMock` in a way to simulate a cache miss. It also looks like this is prevented in https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/Cache/Region/DefaultRegion.php#L89-L91 so theoretically `DefaultMultiGetRegion` can't receive `false` value.

Am I missing something?

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 1]
```diff
@@ -99,7 +99,7 @@ public function get(QueryCacheKey $key, ResultSetMapping $rsm, array $hints = []
         foreach ($cacheEntry->result as $index => $entry) {
             $entityEntry = is_array($entries) ? ($entries[$index] ?? null) : null;
 
-            if ($entityEntry === null) {
+            if (!$entityEntry) {
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: This should rather be `! $entityEntry instanceof CacheEntry` (or something more specific to this usage)

**Comment by lcobucci** [comment]
@umpirsky I think we're looking this from the wrong angle... I mean, if the key exists in Redis it means that it was stored as `false` instead of the expected object and we should look into the write side instead.

**Comment by lcobucci** [comment]
Also, please send the bugfix to `2.6` branch instead of `master`. We'll handle the porting :)

**Comment by umpirsky** [comment]
That's the problem, when I try to create `new CacheEntryMock(false)` I get `InvalidArgumentException: Passed variable is not an array or object`. That's why I can't create failing test.

**Comment by lcobucci** [comment]
@umpirsky I was looking at the write side and had some thoughts that might be related to max value size for strings or serialisation issues... which redis and redis client versions are you using and do you have igbinary installed?

**Comment by lcobucci** [comment]
@umpirsky I just rebased your branch so that we can make this part of the next bugfix release. I'll continue investigating things here.

**Review by lcobucci — DISMISSED** [review]
I'd say things are alright now, although this change doesn't fix the **generation** of such corrupted data (which I couldn't really find :disappointed:).

It would be nice to get another pair of eyes here, just to avoid potential issues.

@umpirsky thanks for your effort and time :+1:

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 2]
```diff
@@ -106,25 +108,27 @@ public function get(QueryCacheKey $key, ResultSetMapping $rsm, array $hints = []
 
         $result      = [];
         $entityName  = reset($rsm->aliasMap);
-        $hasRelation = ( ! empty($rsm->relationMap));
+        $hasRelation = ! empty($rsm->relationMap);
         $persister   = $this->uow->getEntityPersister($entityName);
-        $region      = $persister->getCacheRegion();
-        $regionName  = $region->getName();
+        assert($persister instanceof CachedEntityPersister);
```
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: not super happy with these asserts...

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 3]
```diff
@@ -145,9 +149,11 @@ public function get(QueryCacheKey $key, ResultSetMapping $rsm, array $hints = []
             $data = $entityEntry->data;
 
             foreach ($entry['associations'] as $name => $assoc) {
-                $assocPersister  = $this->uow->getEntityPersister($assoc['targetEntity']);
-                $assocRegion     = $assocPersister->getCacheRegion();
-                $assocMetadata   = $this->em->getClassMetadata($assoc['targetEntity']);
+                $assocPersister = $this->uow->getEntityPersister($assoc['targetEntity']);
+                assert($assocPersister instanceof CachedEntityPersister);
```
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php]: not super happy with these asserts...

**Review by guilhermeblanco — DISMISSED** [review]
Patch seems fine for me aside from the asserts. They should be either removed or try/catch'ed.

**Comment by lcobucci** [comment]
@guilhermeblanco these are meant for development only and replace the not so useful doc blocks for variables.

There's an extensive discussion on that in https://github.com/doctrine/coding-standard/pull/47

**Review by lcobucci — APPROVED** [review]

**Comment by umpirsky** [comment]
Will this be released in `2.6.4` or `2.7`?

**Comment by lcobucci** [comment]
`2.6.4` since it's a bug fix. I plan to release this week.

**Comment by umpirsky** [comment]
Next weekend I guess? :)

**Comment by lcobucci** [comment]
@umpirsky sorry, life happened and didn't allow me to handle things. Will do it ASAP (but won't promise any date :disappointed:)

**Pull request merged**